### ISSUE TITLE
`Rule proposal`: Prefer `Object.entries` and `Object.values` over loop keys

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1066]
## Fail

```js
// Should use `Object.values`
for (const key of Object.keys(foo)) {
	bar(foo[key])
}
```

```js
// Should use `Object.entries`
for (const key of Object.keys(foo)) {
	bar(foo[key], key)
}
```

```js
// Should use `Object.values`
const foo = Object.keys(bar).map(key => baz(bar[key]));
```

```js
// Should use `Object.entries`
const foo = Object.keys(bar).map(key => baz(bar[key], key));
```

## Pass

```js
for (const key of Object.keys(foo)) {
	foo[key] = 'new value'
}
```

**Comment by sindresorhus** [comment]
Rule name: `prefer-object-entries`

**Comment by fisker** [comment]
I think we should also enforce `Object.keys` over `Object.entries` in this case

```js
for (const [key] of Object.entries(foo)) {
	bar(key);
}
```

So better come up a better name cover all them.

**Comment by fisker** [comment]
`proper-object-iterable-methods`?

**Comment by sindresorhus** [comment]
> proper-object-iterable-methods?

Yeah. Sounds ok. I cannot come up with something better.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by ljosberinn** [comment]
a (presumably) more complicated case that could also be detected:
```js
// fail
const foo = Object.keys(myObj).reduce((acc, key) => {
  const value = myObj[key];
  acc[key] = value + 1;
  return acc;
}, {})

// pass
const foo = Object.fromEntries(
  Object.entries(myObj).map(([key, value]) => [key, value + 1])
)
```

**Comment by Pyrolistical** [comment]
While I was doing a refactoring to get rid of a `for..in`. I wondered if this rule should handle this case

```js
// BEFORE
for (const key in object) {
  const value = object[key]
  f(value)
}

// AFTER
for (const [key, value] of Object.entries(object)) {
  f(value)
}
```

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries),
> `Object.entries()` is the same as iterating with a [for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) loop, **except that a `for...in` loop enumerates properties in the prototype chain as well**.

So NO.

**Comment by fregante** [comment]
@Pyrolistical presumably you also have the [`guard-for-in`](https://eslint.org/docs/rules/guard-for-in) rule so:

- the "before" version should not exist
- the gated version should be equivalent and thus fixable

**Comment by Pyrolistical** [comment]
@fregante `for..in` is safe if you know its an object literal. I don't use classes, but you are right this can cause problems to those who do.

```
> for (const key in {x: 42}) { console.log(key) }
x
```

**Comment by fisker** [comment]
> for..in is safe if you know its an object literal.

No, it's not.

```js
Object.prototype.map = () => {}
for (const key in {x: 42}) { console.log(key) }
```

**Comment by Pyrolistical** [comment]
@fisker that's fair. _installs [no-extend-native](https://eslint.org/docs/rules/no-extend-native)_

**Comment by fisker** [comment]
`<script src="lib-add-method-to-object-prototype-but-forgot-set-non-enumeratable.js">`


`import "lib-add-method-to-object-prototype-but-forgot-set-non-enumeratable"`
